lib/solarwinds_apm/opentelemetry/otlp_processor.rb

```diff
@@ -109,31 +119,40 @@ def record_request_metrics(span)
         @metrics[:response_time].record(span_time, attributes: meter_attrs)
       end
 
-      # oboe_api will return 0 in case of failed operation, and report 0 value
-      def record_sampling_metrics
-        _, trace_count = SolarWindsAPM.oboe_api.consumeTraceCount
-        SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] trace_count: #{trace_count}" }
-        @metrics[:tracecount].add(trace_count)
+      # Calculate span time in microseconds (us) using start and end time
+      # in nanoseconds (ns). OTel span start/end_time are optional.
+      def calculate_span_time(start_time: nil, end_time: nil)
+        return 0 if start_time.nil? || end_time.nil?
 
-        _, sample_count = SolarWindsAPM.oboe_api.consumeSampleCount
-        SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] sample_count: #{sample_count}" }
-        @metrics[:samplecount].add(sample_count)
+        ((end_time.to_i - start_time.to_i) / 1e3).round
+      end
 
-        _, request_count = SolarWindsAPM.oboe_api.consumeRequestCount
-        SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] request_count: #{request_count}" }
-        @metrics[:request_count].add(request_count)
+      # Calculate if this span instance has_error
+      # return [Integer]
+      def error?(span)
+        span.status.code == ::OpenTelemetry::Trace::Status::ERROR ? 1 : 0
+      end
 
-        _, token_bucket_exhaustion_count = SolarWindsAPM.oboe_api.consumeTokenBucketExhaustionCount
-        SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] tokenbucket_exhaustion_count: #{token_bucket_exhaustion_count}" }
-        @metrics[:toex_count].add(token_bucket_exhaustion_count)
+      # This span from inbound HTTP request if from a SERVER by some http.method
+      def span_http?(span)
+        span.kind == ::OpenTelemetry::Trace::SpanKind::SERVER && !span.attributes[HTTP_METHOD].nil?
+      end
 
-        _, through_trace_count = SolarWindsAPM.oboe_api.consumeThroughTraceCount
-        SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] through_trace_count: #{through_trace_count}" }
-        @metrics[:through_count].add(through_trace_count)
+      # Calculate HTTP status_code from span or default to UNAVAILABLE
+      # Something went wrong in OTel or instrumented service crashed early
+      # if no status_code in attributes of HTTP span
+      def get_http_status_code(span)
+        span.attributes[HTTP_STATUS_CODE] || INVALID_HTTP_STATUS_CODE
+      end
 
-        _, triggered_trace_count = SolarWindsAPM.oboe_api.consumeTriggeredTraceCount
-        SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] triggered_trace_count: #{triggered_trace_count}" }
-        @metrics[:tt_count].add(triggered_trace_count)
```

Comment:
Are those metrics recorded in the sampler now ?

Reply:
Yes, they are recorded in sampler (initialized from sampling/metrics.rb and used in sampling/oboe_sampler.rb)